H-spaces and homogeneous types

UlrikBuchholtz opened the pull request:
This is a small PR with some connections between H-spaces and homogeneous types.
It also proves that H-space structures can be expressed as a pointed Pi.
It fixes IMO the definition of the pointed identity equivalence.
However, the proof that a homogeneous type can be given an H-space structure is very slow. Any idea what's causing this slowdown?
Once this is faster, I'll add that the corresponding H-space structure is left-invertible.

UlrikBuchholtz commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -71,6 +72,21 @@ HSpace-Π∙-Equiv : {ℓ : Level} (A : Pointed ℓ) → HSpace A ≃
+    k x = pointed-sip⁻ A (typ A , x) (sym (h (pt A)) ∙ h x)
+
+    k₀ : k (pt A) ≡ idEquiv∙ A
+    k₀ = cong′ (pointed-sip⁻ A A) (lCancel (h (pt A))) ∙ pointed-sip⁻-refl A
```
This is the slow line, taking about 20 seconds on my machine to elaborate.

felixwellen replied:
Agda is busy with ```pointed-sip⁻-refl A``` for more than 20 seconds on my machine.

felixwellen replied:
Ah, no, I wasn't careful enough: Only the first 10 seconds or so are spend on the sip-part.

felixwellen replied:
I have no idea yet, why it is so slow. I might get back at it in a couple of hours.

felixwellen replied:
The problem seems to be that ```pointed-sip⁻-refl A``` has a huge normalform. Agda spends 10s on deserializing it and some more time on concatenating it after that.

UlrikBuchholtz replied:
Thanks for looking into this. What's the solution in this case – making it abstract?

felixwellen replied:
I am currently trying, making this particular term abstract only seems to reduce the time somehwat, but it is still >10s.

felixwellen replied:
Aha - making both, pointed-sip-refl and pointed-sip abstract solves it. I'll check if that breaks something...

felixwellen replied:
My network is too bad (train..) to do a PR on your PR, so I just paste here what I did (at the end of Cubical/Homotopy/Pointed.agda):
```agda
pointed-sip-idEquiv∙ : (A : Pointed ℓ) → pointed-sip A A (idEquiv∙ A) ≡ refl
fst (pointed-sip-idEquiv∙ A i j) = uaIdEquiv i j
snd (pointed-sip-idEquiv∙ A i j) = glue {φ = i ∨ ~ j ∨ j} (λ _ → pt A) (pt A)

{-
  The following terms have huge normalforms, so they are abstract to avoid
  type checking speed problems, for example in

    Cubical.Homotopy.HSpace
-}
abstract
  pointed-sip⁻ : (A B : Pointed ℓ) → (A ≡ B) → A ≃[ PointedEquivStr ] B
  pointed-sip⁻ A B = invEq (pointedSIP A B)

  pointed-sip⁻-refl : (A : Pointed ℓ) → pointed-sip⁻ A A refl ≡ idEquiv∙ A
  pointed-sip⁻-refl A = sym (invEq (equivAdjointEquiv (pointedSIP A A)) (pointed-sip-idEquiv∙ A))

pointedEquivAction : EquivAction {ℓ} PointedStructure
pointedEquivAction e = e

pointedTransportStr : TransportStr {ℓ} pointedEquivAction
pointedTransportStr e s = sym (transportRefl _)
```
My (slow) notebook is still checking the library with this change...

felixwellen replied:
Library typechecks.

felixwellen commented:
Ah - now I realized you actually introduced those terms ;-)
I guess your PR is not ready to merge - you can make that visible by converting it to a draft by pressing the "Convert to draft" button/link.

UlrikBuchholtz commented:
Thanks. I added the fact that the H-space structure coming from a homogeneous type is left-invertible as well as a few extra utilities. I'll convert to draft now.

UlrikBuchholtz commented:
Actually, with the new commit I'm happy to have it merged, so I can base more work off of it. Unless there's something I need to fix first?

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
 
+record LeftInvHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ where
+  constructor LeftInvHSp
+  field
```
Just a minor suggestion: You can ```open HSpace e``` here if you like and then use ```μ x``` directly below.
Also, constructors are almost never used since we encourage copattern definitions.

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
+  -- every left-invertible H-space is strongly homogeneous
+  isHomogeneousHSp : isHomogeneous A
+  isHomogeneousHSp x = pointed-sip A (fst A , x)
+                       ((HSpace.μ e x , μ-equiv x) , HSpace.μᵣ e x)
```
```HSpace.μ e``` and ```HSpace.μᵣ e``` *can* be shortened.

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
+                  Π∙ A (λ x → A →∙ (typ A , x)) (idfun∙ A)
+HSpace-Π∙-Equiv A = isoToEquiv (HSpace-Π∙-Iso A)
+
+{- every strongly homogeneous structure on A gives rise
```
I'm a bit confused about the terminology: Is ```isHomogenous``` the same as "strongly homogenous"? Maybe a comment at the definition of ```isHomogenous``` could clarify that.

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
                (μ-assoc (pt A) y z)
                refl
 
+record LeftInvHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ where
```
We put universe levels at the beginning of files (wasn't done in this file yet, so if you don't have the time to fix it, you could just leave it like that) as a private variable (see CONTRIBUTING.md). You can also do that to the second universe level ```you use below and ```A```.

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
+  isHomogeneousHSp x = pointed-sip A (fst A , x)
+                       ((HSpace.μ e x , μ-equiv x) , HSpace.μᵣ e x)
+
+module _ {ℓ ℓ' : Level} {A : Pointed ℓ} {B : Pointed ℓ'} (hb : HSpace B)
```
Universe levels -> variable.

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
+
+{- every strongly homogeneous structure on A gives rise
+   to a left-invertible H-space structure, by a slight modification -}
+HSpace-homogeneous : {ℓ : Level} (A : Pointed ℓ) → isHomogeneous A → HSpace A
```
Universe levels -> variable.

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -27,6 +34,89 @@ record AssocHSpace {ℓ : Level} {A : Pointed ℓ} (e : HSpace A) : Type ℓ whe
+    fst s x = ≃∙map (k x)
+    snd s = cong′ ≃∙map k₀
+
+LeftInvHSpace-homogeneous : {ℓ : Level} (A : Pointed ℓ) → (h : isHomogeneous A)
```
Universe levels -> variable.

UlrikBuchholtz commented:
Thanks for the comments! I've fixed the level variables and such, and I've added a comment to Cubical.Foundations.Pointed.Homogeneous to clarify that it's quite a bit stronger than what might be the expected notion. I hope that's clarifying!
In commit 0df1c651 I also prove that an H-space structure on a codomain gives rise to an H-space structure on a pointed mapping type. (This is involves the first hypercube I've done in cubical, yay!)

felixwellen commented on Cubical/Homotopy/HSpace.agda:
```diff
@@ -49,3 +143,59 @@ S1-AssocHSpace : AssocHSpace S1-HSpace
 μ-assoc-filler S1-AssocHSpace _ _ = refl
+
+-- →∙
+→∙-HSpace : (A : Pointed ℓ) (X : Pointed ℓ') (hX : HSpace X) → HSpace (A →∙ X ∙)
```
If you like, you can introduce an anonymous module with the arguments 
```(A : Pointed ℓ) (X : Pointed ℓ') (hX : HSpace X)```
Then you could open hX and have a less noisy pattern matching.

UlrikBuchholtz replied:
Thanks, done in d80a3b4a

felixwellen commented:
I wrote one suggestions you might wan't to implement...
But if you don't, just let me know, then I'll merge it.
And you are now one dimension ahead of me cubewise ;-)

felixwellen commented:
Looks all good to me, I'll merge.
